File: python/medousa-sdk/src/medousa/bots.py
from __future__ import annotations

from medousa._decode import decode
from medousa._ops import op_path
from medousa.client import MedousaClient
from medousa.types import (
    BotListResponse,
    BotOpenResponse,
    BotProfile,
    CreateBotRequest,
    DuplicateBotRequest,
    SessionBotResponse,
    SetBotArchivedRequest,
    SetSessionBotRequest,
    UpdateBotRequest,
)


class BotsApi:
    def __init__(self, client: MedousaClient) -> None:
        self._client = client

    async def list(self) -> BotListResponse:
        value = await self._client.transport.get_json(
            self._client.base_url, op_path("bots.get")
        )
        return decode(BotListResponse, value)

    async def create(self, request: CreateBotRequest) -> BotOpenResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("bots.post"),
            request.model_dump(mode="json"),
        )
        return decode(BotOpenResponse, value)

    async def get(self, bot_id: str) -> BotProfile:
        value = await self._client.transport.get_json(
            self._client.base_url, op_path("bots.by_bot_id.get", bot_id=bot_id)
        )
        return decode(BotProfile, value)

    async def update(self, bot_id: str, request: UpdateBotRequest) -> BotProfile:
        value = await self._client.transport.put_json(
            self._client.base_url,
            op_path("bots.by_bot_id.put", bot_id=bot_id),
            request.model_dump(mode="json"),
        )
        return decode(BotProfile, value)

    async def set_archived(
        self, bot_id: str, request: SetBotArchivedRequest
    ) -> BotProfile:
        value = await self._client.transport.put_json(
            self._client.base_url,
            op_path("bots.by_bot_id.archive.put", bot_id=bot_id),
            request.model_dump(mode="json"),
        )
        return decode(BotProfile, value)

    async def duplicate(
        self, bot_id: str, request: DuplicateBotRequest
    ) -> BotOpenResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("bots.by_bot_id.duplicate.post", bot_id=bot_id),
            request.model_dump(mode="json"),
        )
        return decode(BotOpenResponse, value)

    async def open(self, bot_id: str) -> BotOpenResponse:
        value = await self._client.transport.post_empty_json(
            self._client.base_url,
            op_path("bots.by_bot_id.open.post", bot_id=bot_id),
        )
        return decode(BotOpenResponse, value)

    async def session(self, session_id: str) -> SessionBotResponse:
        value = await self._client.transport.get_json(
            self._client.base_url,
            op_path("sessions.by_session_id.bot.get", session_id=session_id),
        )
        return decode(SessionBotResponse, value)

    async def bind_session(
        self, session_id: str, request: SetSessionBotRequest
    ) -> SessionBotResponse:
        value = await self._client.transport.put_json(
            self._client.base_url,
            op_path("sessions.by_session_id.bot.put", session_id=session_id),
            request.model_dump(mode="json"),
        )
        return decode(SessionBotResponse, value)

    async def unbind_session(self, session_id: str) -> SessionBotResponse:
        value = await self._client.transport.delete_json(
            self._client.base_url,
            op_path("sessions.by_session_id.bot.delete", session_id=session_id),
        )
        return decode(SessionBotResponse, value)

File: crates/medousa-sdk/src/bots.rs
#[cfg(feature = "async")]
use medousa_types::{
    BotListResponse, BotOpenResponse, BotProfile, CreateBotRequest, DuplicateBotRequest,
    SessionBotResponse, SetBotArchivedRequest, SetSessionBotRequest, UpdateBotRequest,
};

#[cfg(feature = "async")]
use crate::client::MedousaClient;
use crate::generated::ops;
use crate::op::op_path;
#[cfg(feature = "async")]
use crate::transport::decode;

#[cfg(feature = "async")]
pub struct BotsApi<'a> {
    pub(crate) client: &'a MedousaClient,
}

#[cfg(feature = "async")]
impl BotsApi<'_> {
    pub async fn list(&self) -> Result<BotListResponse, crate::SdkError> {
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), ops::BOTS_GET.path)
            .await?;
        decode(value).await
    }

    pub async fn create(
        &self,
        request: &CreateBotRequest,
    ) -> Result<BotOpenResponse, crate::SdkError> {
        let body = serde_json::to_value(request)
            .map_err(|error| crate::SdkError::Serde(error.to_string()))?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), ops::BOTS_POST.path, body)
            .await?;
        decode(value).await
    }

    pub async fn get(&self, bot_id: &str) -> Result<BotProfile, crate::SdkError> {
        let path = op_path(&ops::BOTS_BY_BOT_ID_GET, &[("bot_id", bot_id)])?;
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }

    pub async fn update(
        &self,
        bot_id: &str,
        request: &UpdateBotRequest,
    ) -> Result<BotProfile, crate::SdkError> {
        let path = op_path(&ops::BOTS_BY_BOT_ID_PUT, &[("bot_id", bot_id)])?;
        let body = serde_json::to_value(request)
            .map_err(|error| crate::SdkError::Serde(error.to_string()))?;
        let value = self
            .client
            .transport()
            .put_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn set_archived(
        &self,
        bot_id: &str,
        request: &SetBotArchivedRequest,
    ) -> Result<BotProfile, crate::SdkError> {
        let path = op_path(&ops::BOTS_BY_BOT_ID_ARCHIVE_PUT, &[("bot_id", bot_id)])?;
        let body = serde_json::to_value(request)
            .map_err(|error| crate::SdkError::Serde(error.to_string()))?;
        let value = self
            .client
            .transport()
            .put_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn duplicate(
        &self,
        bot_id: &str,
        request: &DuplicateBotRequest,
    ) -> Result<BotOpenResponse, crate::SdkError> {
        let path = op_path(&ops::BOTS_BY_BOT_ID_DUPLICATE_POST, &[("bot_id", bot_id)])?;
        let body = serde_json::to_value(request)
            .map_err(|error| crate::SdkError::Serde(error.to_string()))?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn open(&self, bot_id: &str) -> Result<BotOpenResponse, crate::SdkError> {
        let path = op_path(&ops::BOTS_BY_BOT_ID_OPEN_POST, &[("bot_id", bot_id)])?;
        let value = self
            .client
            .transport()
            .post_empty_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }

    pub async fn session(&self, session_id: &str) -> Result<SessionBotResponse, crate::SdkError> {
        let path = op_path(
            &ops::SESSIONS_BY_SESSION_ID_BOT_GET,
            &[("session_id", session_id)],
        )?;
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }

    pub async fn bind_session(
        &self,
        session_id: &str,
        request: &SetSessionBotRequest,
    ) -> Result<SessionBotResponse, crate::SdkError> {
        let path = op_path(
            &ops::SESSIONS_BY_SESSION_ID_BOT_PUT,
            &[("session_id", session_id)],
        )?;
        let body = serde_json::to_value(request)
            .map_err(|error| crate::SdkError::Serde(error.to_string()))?;
        let value = self
            .client
            .transport()
            .put_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn unbind_session(
        &self,
        session_id: &str,
    ) -> Result<SessionBotResponse, crate::SdkError> {
        let path = op_path(
            &ops::SESSIONS_BY_SESSION_ID_BOT_DELETE,
            &[("session_id", session_id)],
        )?;
        let value = self
            .client
            .transport()
            .delete_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }
}
